go-controller/pkg/libovsdbops/meter.go

```diff
@@ -1,118 +1,63 @@
 package libovsdbops
 
 import (
-	"context"
-	"fmt"
+	"reflect"
 
 	libovsdbclient "github.com/ovn-org/libovsdb/client"
 	"github.com/ovn-org/libovsdb/ovsdb"
+	"k8s.io/apimachinery/pkg/util/sets"
 
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
-	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
 )
 
-// FindMeterByName finds an ovn meter by name, will return `errNotFound` if it does not exist
-func FindMeterByName(nbClient libovsdbclient.Client, name string) (*nbdb.Meter, error) {
-	searched := &nbdb.Meter{
-		Name: name,
-	}
-
-	ctx, cancel := context.WithTimeout(context.Background(), types.OVSDBTimeout)
-	defer cancel()
-	err := nbClient.Get(ctx, searched)
-	if err != nil {
-		return nil, err
-	}
-
-	return searched, nil
+func equalsMeterBand(a, b *nbdb.MeterBand) bool {
+	return a.Action == b.Action &&
+		a.BurstSize == b.BurstSize &&
+		a.Rate == b.Rate &&
+		reflect.DeepEqual(a.ExternalIDs, b.ExternalIDs)
 }
 
-// UpdateMeterFairness updates the `fair` column of an ovn Meter
-func UpdateMeterFairness(nbClient libovsdbclient.Client, meter *nbdb.Meter, fairness bool) error {
-	meter.Fair = &fairness
-
+// CreateMeterBandOps creates the provided meter band if it does not exist
+func CreateMeterBandOps(nbClient libovsdbclient.Client, ops []ovsdb.Operation, meterBand *nbdb.MeterBand) ([]ovsdb.Operation, error) {
+	bands := []*nbdb.MeterBand{}
 	opModel := OperationModel{
-		Model: meter,
-		OnModelUpdates: []interface{}{
-			&meter.Fair,
-		},
-		ErrNotFound: true,
-	}
-
-	m := NewModelClient(nbClient)
-	if _, err := m.CreateOrUpdate(opModel); err != nil {
-		return fmt.Errorf("error while updating Meter Fairness to: %v error %v", fairness, err)
-	}
-
-	return nil
-}
-
-// CreateMeterWithBand simulates the ovn-nbctl operation performed by the `meter-add` command. i.e create the Meter
-// and accompanying meter_band, and add the meterband to the meter only call this if the meter does not exist
-func CreateMeterWithBand(nbClient libovsdbclient.Client, meter *nbdb.Meter, meterBand *nbdb.MeterBand) ([]ovsdb.OperationResult, error) {
-	opModels := []OperationModel{
-		{
-			Model: meterBand,
-			DoAfter: func() {
-				meter.Bands = []string{meterBand.UUID}
-			},
-		},
-		{
-			Model: meter,
+		Model:          meterBand,
+		ModelPredicate: func(item *nbdb.MeterBand) bool { return equalsMeterBand(item, meterBand) },
+		OnModelUpdates: nil, // no update
+		ExistingResult: &bands,
+		DoAfter: func() {
+			// in case we have multiple equal bands, pick the first one for
```

Comment:
hmm if we include the UUID as an equal it should be fine right? should we be having duplicate meter bands with exact same action/burst/rate/externalids?

Reply:
We can't include the UUID because we might not have it. If we do, the predicate is not used anyway.

Before, we were creating duplicate bands for different meters. Now, we don't, if two meter need the same band we just have that one band. Relying on the server to remove unused bands did not change, that was already happening.

Reply:
oh yes, @flavio-fernandes and I did a dirty hack to make sure we didn't have to create duplicate bands,  I get your point. thanks for clarifying!